patches/net/minecraft/world/level/chunk/ChunkAccess.java.patch

```diff
@@ -9,35 +9,35 @@
      public static final int NO_FILLED_SECTION = -1;
      private static final Logger LOGGER = LogUtils.getLogger();
      private static final LongSet EMPTY_REFERENCE_SET = new LongOpenHashSet();
-@@ -308,23 +_,28 @@
+@@ -308,10 +_,19 @@
  
      @Override
      public final void findBlockLightSources(BiConsumer<BlockPos, BlockState> p_285269_) {
 -        this.findBlocks(p_284897_ -> p_284897_.getLightEmission() != 0, p_285269_);
-+        this.findBlocks((p_284897_, pos) -> p_284897_.getLightEmission(this, pos) != 0, p_285269_);
++        this.findBlocks(p_284897_ -> p_284897_.hasDynamicLightEmission() || p_284897_.getLightEmission(net.minecraft.world.level.EmptyBlockGetter.INSTANCE, BlockPos.ZERO) != 0, (p_284897_, pos) -> p_284897_.getLightEmission(this, pos) != 0, p_285269_);
      }
  
      public void findBlocks(Predicate<BlockState> p_285343_, BiConsumer<BlockPos, BlockState> p_285030_) {
-+        findBlocks((state, pos) -> p_285343_.test(state), p_285030_);
++        findBlocks(p_285343_, (state, pos) -> p_285343_.test(state), p_285030_);
 +    }
 +
++    @Deprecated(forRemoval = true)
 +    public void findBlocks(java.util.function.BiPredicate<BlockState, BlockPos> p_285343_, BiConsumer<BlockPos, BlockState> p_285030_) {
++        findBlocks(state -> p_285343_.test(state, BlockPos.ZERO), p_285343_, p_285030_);
++    }
++
++    public void findBlocks(Predicate<BlockState> p_285343_, java.util.function.BiPredicate<BlockState, BlockPos> fineFilter, BiConsumer<BlockPos, BlockState> p_285030_) {
          BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
  
          for(int i = this.getMinSection(); i < this.getMaxSection(); ++i) {
-             LevelChunkSection levelchunksection = this.getSection(this.getSectionIndexFromSectionY(i));
--            if (levelchunksection.maybeHas(p_285343_)) {
-+            if (levelchunksection.maybeHas((state) -> p_285343_.test(state, BlockPos.ZERO))) {
-                 BlockPos blockpos = SectionPos.of(this.chunkPos, i).origin();
- 
-                 for(int j = 0; j < 16; ++j) {
+@@ -323,8 +_,9 @@
                      for(int k = 0; k < 16; ++k) {
                          for(int l = 0; l < 16; ++l) {
                              BlockState blockstate = levelchunksection.getBlockState(l, j, k);
 -                            if (p_285343_.test(blockstate)) {
 -                                p_285030_.accept(blockpos$mutableblockpos.setWithOffset(blockpos, l, j, k), blockstate);
 +                            blockpos$mutableblockpos.setWithOffset(blockpos, l, j, k);
-+                            if (p_285343_.test(blockstate, blockpos$mutableblockpos.immutable())) {
++                            if (fineFilter.test(blockstate, blockpos$mutableblockpos.immutable())) {
```

Comment:
I realize the original patch also did this, but using a mutable pos is pointless if `immutable` is called on every iteration of the loop (as `immutable` will always allocate a new immutable BlockPos).

Reply:
Unless anyone can think of a reason to not do it, I'd be fine with removing the conversion to an immutable `BlockPos` since this would only call getters where the danger of the position being stored and the mutability being leaked is very low.

Reply:
There are a lot of vanilla systems which use `MutableBlockPos` but don't expose it (performance-sensitive code paths such as entity pathing, entity collision, lighting, worldgen and rendering), so expecting an immutable position in getters related to these systems is not a great idea.

Also, to put this into perspective: in the worst case of every chunk section in a given chunk having at least one (potentially) light-emitting block, this code is currently allocating 98.3k `BlockPos` instances per chunk (4096 instances per section at 24 sections in case of default world height) instead of a single one. Is that really worth it for a tiny bit of idiot-proofing?

Reply:
Meh you are right.